aspnetcore/performance/rate-limit-samples.md

```diff
@@ -0,0 +1,58 @@
+---
+title: Rate limiting middleware samsples
+author: rick-anderson
+ms.author: riande
+monikerRange: '>= aspnetcore-7.0'
+description: Samples for using ASP.NET rate limitng middleware
+ms.custom: mvc
+ms.date: 03/05/2025
+uid: performance/rate-limit
+---
+
+# Rate limiter samples
+
+The following samples aren't meant for production code but are examples on how to use the limiters.
+
+### Limiter with `OnRejected`, `RetryAfter`, and `GlobalLimiter`
+
+The following sample:
+
+* Creates a [RateLimiterOptions.OnRejected](xref:Microsoft.AspNetCore.RateLimiting.RateLimiterOptions.OnRejected) callback that is called when a request exceeds the specified limit. `retryAfter` can be used with the [TokenBucketRateLimiter](/dotnet/api/system.threading.ratelimiting.tokenbucketratelimiter), [FixedWindowLimiter](/dotnet/api/microsoft.aspnetcore.ratelimiting.ratelimiteroptionsextensions.addfixedwindowlimiter), and [SlidingWindowLimiter](/dotnet/api/microsoft.aspnetcore.ratelimiting.ratelimiteroptionsextensions.addslidingwindowlimiter) because these algorithms are able to estimate when more permits will be added. The `ConcurrencyLimiter` has no way of calculating when permits will be available.
+* Adds the following limiters:
+
+  * A `SampleRateLimiterPolicy` which implements the `IRateLimiterPolicy<TPartitionKey>` interface. The `SampleRateLimiterPolicy` class is shown later in this article.
+  * A `SlidingWindowLimiter`:
+    * With a partition for each authenticated user.
+    * One shared partition for all anonymous users.
+  * A <xref:Microsoft.AspNetCore.RateLimiting.RateLimiterOptions.GlobalLimiter> that is applied to all requests. The global limiter will be executed first, followed by the endpoint-specific limiter, if one exists. The `GlobalLimiter` creates a partition for each <xref:System.Net.IPAddress>.
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs" id="snippet_1":::
+
+> [!WARNING]
+>Creating partitions on client IP addresses makes the app vulnerable to Denial of Service Attacks which employ IP Source Address Spoofing. For more information, see [BCP 38 RFC 2827 Network Ingress Filtering: Defeating Denial of Service Attacks which employ IP Source Address Spoofing](https://www.rfc-editor.org/info/bcp38).
+
+See [the samples repository for the complete `Program.cs`](https://github.com/dotnet/AspNetCore.Docs.Samples/blob/main/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs#L145,L281) file.
```

Comment:
Note that the line referencing approach "`#L145,L281`" won't work. Perhaps best to let the dev find the lines, especially given that the PU likes to move lines around.

```suggestion
For the complete `Program.cs` file, see [the samples repository](https://github.com/dotnet/AspNetCore.Docs.Samples/blob/main/fundamentals/middleware/rate-limit/WebRateLimitAuth/Program.cs).

[!INCLUDE[](~/includes/aspnetcore-repo-ref-source-links.md)]
```